libuavcan/include/libuavcan/platform/memory.hpp

```diff
@@ -0,0 +1,313 @@
+/*
+ * Copyright (C) 2014 Pavel Kirienko <pavel.kirienko@gmail.com>
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ */
+#ifndef LIBUAVCAN_PLATFORM_MEMORY_HPP_INCLUDED
+#define LIBUAVCAN_PLATFORM_MEMORY_HPP_INCLUDED
+
+#include <atomic>
+#include <unordered_map>
+#include "libuavcan/libuavcan.hpp"
+#include "libuavcan/introspection.hpp"
+
+namespace libuavcan
+{
+namespace platform
+{
+namespace memory
+{
+/**
+ * Classic implementation of a pool allocator (Meyers).
+ *
+ * This instance is designed to be a process-wide singleton and is not designed for direct access
+ * by applications or libuavcan. To use this memory pool it is recommended that one of the allocators
+ * defined in this header is used.
+ *
+ * This implementation relies on STL atomics and thread-safe static compiler support to be thread-safe.
+ * If compiling with -fno-threadsafe-statics then the application must ensure no concurrent access is
+ * attempted the first time getInstance() is invoked. After this first access all subsequent access
+ * is thread-safe regardless of thread-safe static support.
+ *
+ * @tparam  NumBlocksParam  The number of blocks to allocate in the pool. Unlike an arena allocator,
+ *                          this allocator does not guarantee a continuous block of memory and each
+ *                          memory block can located anywhere on the system. It is highly recommended
+ *                          that all memory blocks have the same access permsissions and performance.
+ * @tparam  BlockSizeParam  The size in bytes of each block. Note that all blocks will use this size
+ *                          such that the amount of memory allocated for the pool will be at least
+ *                          NumBlocksParam x BlockSizeParam bytes. Implementations may choose to
+ *                          allocate additional memory per-block to detect buffer overrun and illegal
+ *                          deallocation.
+ */
+template <std::size_t NumBlocksParam, std::uint8_t BlockSizeParam>
+class LIBUAVCAN_EXPORT StaticMemoryPool final
```

Comment:
PX4 is extremely memory-limited so we had to implement a high-watermark allocator backed by the heap: when allocating, check the cache and use heap if empty; when deallocating, put the reclaimed block into the cache. I expect that to be still needed in PX4 (and similar systems) because worst-case memory usage depends on the configuration of the UAVCAN stack and we can't recompile the firmware when its configuration is changed.

Reply:
I'm unsure what  actions to take here based on this comment. This design will allow for different allocators to be utilized with libuavcan as long as they support STL allocator concepts. I'm porting the  v0 memory  pool allocator and then STL comes with its own heap allocator so are you also asking for a "high-watermark" allocator to be included with v1? If so, that seems like something to do in another change. Or are you saying we should get rid of the pool allocator? That'd be problematic for systems that do not  have a heap.

Reply:
> are you also asking for a "high-watermark" allocator to be included with v1? If so, that seems like something to do in another change

Consensus here.

Reply:
Added #263 . Please provide requirements and details in that issue.